### PR TITLE
Utils: shortened brew paths

### PR DESCRIPTION
**Pull request by alebcay**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

### Overview

This PR adds:
- New file in Utils module, `utils/path`
  - `Utils.path_is_parent_of?` determines if one `Pathname`/`String` is a parent of another `Pathname`/`String`
  - `Utils.shortened_brew_path` expresses a `Pathname`/`String` as a shell expansion where possible, e.g. `/usr/local/Cellar/cmake` as `$(brew --cellar)/cmake`. Where possible, it can also express a `Pathname` as a formula's opt prefix, e.g. `/usr/local/opt/openjdk/libexec/openjdk.jdk/Contents/Home` becomes `$(brew --prefix openjdk)/libexec/openjdk.jdk/Contents/Home`
- Methods in `language/java`
  - `Language::Java.short_java_home_shell`, `Language::Java.short_java_home_env`, `Language::Java.overridable_short_java_home_env` which are shortened versions of the existing respective methods

This PR also modifies `write_env_script`, `write_exec_script`, and `write_jar_script` to use `shortened_brew_path`.

### Rationale

Now that we are bottling everything, there are many cases where we can't use `all:` bottles because an env script, exec script, jar script etc. has a platform-dependent `HOMEBREW_PREFIX` in it (e.g. `/usr/local` vs `/opt/homebrew`).

The result is that we have to carry extra copies of some bottles where the bulk of the payload is the same. This is especially the case in Java based formulae, where the resulting bottle can be tens or hundreds of megabytes in size.

Commands like `$(brew --prefix)`, `$(brew --cellar)`, etc. don't invoke the full Homebrew installation (they short-circuit to shell script instead), so using these where possible in scripts makes the resulting bottle portable while adding minimal performance cost.

A resulting env script might look something like this:
```bash
#!/bin/bash
JAVA_HOME="${JAVA_HOME:-$(brew --prefix openjdk)/libexec/openjdk.jdk/Contents/Home}" exec "$(brew --cellar)/mill/0.9.8/libexec/mill"  "$@"
```

I'm especially open to input on the naming of any of the new methods introduced in this PR. Wasn't really sure about a most suitable name so just went with my gut for now.

**Note**: I'm not happy with the tests for the additional Java methods right now. In the test env, they evaluate to `$(brew --prefix)/opt/openjdk/libexec/...` when in practice (from testing with formulae) it actually produces a more desired `$(brew --prefix openjdk)/libexec/...`. Both are technically correct (when `openjdk` is installed), but just wanted to point out this slight discrepancy. I think this is due to how the `openjdk` formula is being mocked in that test, but after trying to manually set up folders/symlink them where appropriate in the test, I still couldn't get it working exactly as desired.

### CODE REVIEW

File: Library/Homebrew/extend/pathname.rb [hunk 1]
```diff
@@ -432,6 +432,47 @@ def install_metafiles(from = Pathname.pwd)
     end
   end
 
+  sig { params(other_path: T.any(String, Pathname)).returns(T::Boolean) }
+  def parent_of?(other_path)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/extend/pathname.rb:436]: I'd rather we didn't add more methods on `extend/Pathname`. Ideally I think we should remove all pathname extensions (or at least those that are there by default and not added later like e.g. `GitRepository`)

**Review by MikeMcQuaid — COMMENTED** [review]
I like the idea but implementation needs some tweaks!

File: Library/Homebrew/extend/pathname.rb (hunk 1 above)
Comment (alebcay) [reply Library/Homebrew/extend/pathname.rb:436]: Fair enough. Should I leave `parent_of?` here and have `shortened_brew_path` live in e.g. `Utils`? Or move both to `Utils`?

Comment (MikeMcQuaid) [reply Library/Homebrew/extend/pathname.rb:436]: Ideally both move to utils, I think.

**Comment by alebcay** [comment]
Moved some things around. Now the new methods live in `utils/path`.

**Review by MikeMcQuaid — APPROVED** [review]
Great work! 🎉

**Pull request merged**